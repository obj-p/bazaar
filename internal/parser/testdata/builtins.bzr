package builtins

template BuiltinFunctions(list [string], items [int]) {
    // Basic built-in functions
    for i in range(0, 10) {}
    for (idx, val) in enumerate(list) {}
    var size = len(items)

    // Complex expressions with built-ins
    if len(items) > 0 {}
    for i in range(0, len(list)) {}

    // Built-ins in various contexts
    var rangeVar = range(1, 5)
    var enumerateVar = enumerate(items)
    var lengthCheck = len(list) + len(items)

    // Nested built-in calls
    for (i, item) in enumerate(range(0, len(list))) {}

    // Built-ins with complex arguments
    for entry in enumerate(list) {}
    var complexRange = range(len(items), len(items) + 10)
}
