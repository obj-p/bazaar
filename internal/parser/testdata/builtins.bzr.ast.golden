&parser.Bazaar{
  Package: &parser.PackageDecl{
    Domain: []string{
      "builtins",
    },
  },
  Decls: []*parser.TopLevelDecl{
    {
      Function: &parser.FunctionDecl{
        Name: "BuiltinFunctions",
        Parameters: []*parser.ParameterDecl{
          {
            Name: "list",
            Type: &parser.TypeDecl{
              Array: &parser.ArrayTypeDecl{
                Value: parser.TypeDecl{
                  Value: &"string",
                },
              },
            },
          },
          {
            Name: "items",
            Type: &parser.TypeDecl{
              Array: &parser.ArrayTypeDecl{
                Value: parser.TypeDecl{
                  Value: &"int",
                },
              },
            },
          },
        },
        Block: []*parser.Stmt{
          {
            For: &parser.ForStmt{
              Dest: &parser.VariableExpr{
                Name: &"i",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"range",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Literal: &parser.Literal{
                                      Number: &"0",
                                    },
                                  },
                                },
                              },
                            },
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Literal: &parser.Literal{
                                      Number: &"10",
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            For: &parser.ForStmt{
              Dest: &parser.VariableExpr{
                Destructuring: &parser.DestructuringExpr{
                  Names: []string{
                    "idx",
                    "val",
                  },
                },
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"enumerate",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Reference: &parser.ReferenceExpr{
                                      Name: &"list",
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"size",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"len",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Reference: &parser.ReferenceExpr{
                                      Name: &"items",
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            If: &parser.IfStmt{
              Fragments: []*parser.IfFragment{
                {
                  Expr: &parser.Expr{
                    Binary: &parser.BinaryExpr{
                      Left: &parser.Expr{
                        Unary: &parser.UnaryExpr{
                          Primary: &parser.PrimaryExpr{
                            Reference: &parser.ReferenceExpr{
                              Name: &"len",
                              KeyPath: &parser.KeyPathExpr{
                                Call: &parser.CallExpr{
                                  Arguments: []*parser.ArgumentExpr{
                                    {
                                      Expr: &parser.Expr{
                                        Unary: &parser.UnaryExpr{
                                          Primary: &parser.PrimaryExpr{
                                            Reference: &parser.ReferenceExpr{
                                              Name: &"items",
                                            },
                                          },
                                        },
                                      },
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                      Op: token.Op(token.OpGt),
                      Right: &parser.Expr{
                        Unary: &parser.UnaryExpr{
                          Primary: &parser.PrimaryExpr{
                            Literal: &parser.Literal{
                              Number: &"0",
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            For: &parser.ForStmt{
              Dest: &parser.VariableExpr{
                Name: &"i",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"range",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Literal: &parser.Literal{
                                      Number: &"0",
                                    },
                                  },
                                },
                              },
                            },
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Reference: &parser.ReferenceExpr{
                                      Name: &"len",
                                      KeyPath: &parser.KeyPathExpr{
                                        Call: &parser.CallExpr{
                                          Arguments: []*parser.ArgumentExpr{
                                            {
                                              Expr: &parser.Expr{
                                                Unary: &parser.UnaryExpr{
                                                  Primary: &parser.PrimaryExpr{
                                                    Reference: &parser.ReferenceExpr{
                                                      Name: &"list",
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                      },
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"rangeVar",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"range",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Literal: &parser.Literal{
                                      Number: &"1",
                                    },
                                  },
                                },
                              },
                            },
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Literal: &parser.Literal{
                                      Number: &"5",
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"enumerateVar",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"enumerate",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Reference: &parser.ReferenceExpr{
                                      Name: &"items",
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"lengthCheck",
              },
              Source: &parser.Expr{
                Binary: &parser.BinaryExpr{
                  Left: &parser.Expr{
                    Unary: &parser.UnaryExpr{
                      Primary: &parser.PrimaryExpr{
                        Reference: &parser.ReferenceExpr{
                          Name: &"len",
                          KeyPath: &parser.KeyPathExpr{
                            Call: &parser.CallExpr{
                              Arguments: []*parser.ArgumentExpr{
                                {
                                  Expr: &parser.Expr{
                                    Unary: &parser.UnaryExpr{
                                      Primary: &parser.PrimaryExpr{
                                        Reference: &parser.ReferenceExpr{
                                          Name: &"list",
                                        },
                                      },
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                  Op: token.Op(token.OpAdd),
                  Right: &parser.Expr{
                    Unary: &parser.UnaryExpr{
                      Primary: &parser.PrimaryExpr{
                        Reference: &parser.ReferenceExpr{
                          Name: &"len",
                          KeyPath: &parser.KeyPathExpr{
                            Call: &parser.CallExpr{
                              Arguments: []*parser.ArgumentExpr{
                                {
                                  Expr: &parser.Expr{
                                    Unary: &parser.UnaryExpr{
                                      Primary: &parser.PrimaryExpr{
                                        Reference: &parser.ReferenceExpr{
                                          Name: &"items",
                                        },
                                      },
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            For: &parser.ForStmt{
              Dest: &parser.VariableExpr{
                Destructuring: &parser.DestructuringExpr{
                  Names: []string{
                    "i",
                    "item",
                  },
                },
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"enumerate",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Reference: &parser.ReferenceExpr{
                                      Name: &"range",
                                      KeyPath: &parser.KeyPathExpr{
                                        Call: &parser.CallExpr{
                                          Arguments: []*parser.ArgumentExpr{
                                            {
                                              Expr: &parser.Expr{
                                                Unary: &parser.UnaryExpr{
                                                  Primary: &parser.PrimaryExpr{
                                                    Literal: &parser.Literal{
                                                      Number: &"0",
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                            {
                                              Expr: &parser.Expr{
                                                Unary: &parser.UnaryExpr{
                                                  Primary: &parser.PrimaryExpr{
                                                    Reference: &parser.ReferenceExpr{
                                                      Name: &"len",
                                                      KeyPath: &parser.KeyPathExpr{
                                                        Call: &parser.CallExpr{
                                                          Arguments: []*parser.ArgumentExpr{
                                                            {
                                                              Expr: &parser.Expr{
                                                                Unary: &parser.UnaryExpr{
                                                                  Primary: &parser.PrimaryExpr{
                                                                    Reference: &parser.ReferenceExpr{
                                                                      Name: &"list",
                                                                    },
                                                                  },
                                                                },
                                                              },
                                                            },
                                                          },
                                                        },
                                                      },
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                      },
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            For: &parser.ForStmt{
              Dest: &parser.VariableExpr{
                Name: &"entry",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"enumerate",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Reference: &parser.ReferenceExpr{
                                      Name: &"list",
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"complexRange",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Reference: &parser.ReferenceExpr{
                      Name: &"range",
                      KeyPath: &parser.KeyPathExpr{
                        Call: &parser.CallExpr{
                          Arguments: []*parser.ArgumentExpr{
                            {
                              Expr: &parser.Expr{
                                Unary: &parser.UnaryExpr{
                                  Primary: &parser.PrimaryExpr{
                                    Reference: &parser.ReferenceExpr{
                                      Name: &"len",
                                      KeyPath: &parser.KeyPathExpr{
                                        Call: &parser.CallExpr{
                                          Arguments: []*parser.ArgumentExpr{
                                            {
                                              Expr: &parser.Expr{
                                                Unary: &parser.UnaryExpr{
                                                  Primary: &parser.PrimaryExpr{
                                                    Reference: &parser.ReferenceExpr{
                                                      Name: &"items",
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                      },
                                    },
                                  },
                                },
                              },
                            },
                            {
                              Expr: &parser.Expr{
                                Binary: &parser.BinaryExpr{
                                  Left: &parser.Expr{
                                    Unary: &parser.UnaryExpr{
                                      Primary: &parser.PrimaryExpr{
                                        Reference: &parser.ReferenceExpr{
                                          Name: &"len",
                                          KeyPath: &parser.KeyPathExpr{
                                            Call: &parser.CallExpr{
                                              Arguments: []*parser.ArgumentExpr{
                                                {
                                                  Expr: &parser.Expr{
                                                    Unary: &parser.UnaryExpr{
                                                      Primary: &parser.PrimaryExpr{
                                                        Reference: &parser.ReferenceExpr{
                                                          Name: &"items",
                                                        },
                                                      },
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                      },
                                    },
                                  },
                                  Op: token.Op(token.OpAdd),
                                  Right: &parser.Expr{
                                    Unary: &parser.UnaryExpr{
                                      Primary: &parser.PrimaryExpr{
                                        Literal: &parser.Literal{
                                          Number: &"10",
                                        },
                                      },
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
}
